### DISCUSSION TITLE
midi map mode

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#7] [General]
According to https://nsuspray.github.io/Live_API_Doc/8.2.7-10.1.0.xml  midimapmode has changed between Live8 and Live10 like this:
map_midi_note_with_feedback_map()Built-In

map_midi_note_with_feedback_map( (int)arg1, (DeviceParameter)arg2, (int)arg3, (int)arg4, (object)arg5) → bool

map_midi_note_with_feedback_map( (int)arg1, (DeviceParameter)arg2, (int)arg3, (int)arg4, (NoteFeedbackRule)arg5) → bool

So the only change I see here is that arg5 was an object and is now a ? maybe that's the boost problem?

**Comment by markusschloesser** [comment]
@BilldarBagdar  I've compared the XML/stub again and the difference now is (in terms of what the c++ signature expects):

I had to reformat to include the python stuff and what is actually send:
```
(int)midi_map_handle	param					0			encoder			Live.MidiMap.MapMode.relative_signed_bit	feedback_rule					needs_takeover		sensitivity=1.0
(int)			(DeviceParameter)			(int)midi_channel	(int)controller_number	(MapMode)					(CCFeedbackRule)				(bool)avoid_takeover	(float)sensitivity=1.0]) -&gt; bool
unsigned long		class TPyHandle&lt;class ATimeable&gt;	long			long			enum NRemoteMapperTypes::TControllerMapMode	class NPythonMidiMap::TCCFeedbackRule		bool 			[,float=1.0])
unsigned int		TPyHandle&lt;ATimeableValue&gt;		int			int			NRemoteMapperTypes::TControllerMapMode		NPythonMidiMap::TCCFeedbackRule			bool 			[,float=1.0])
```
Line 3 being the old, Line 4 being Live 11/py3.
1. When `TPyHandle`,  `ATimeable` and `NPythonMidiMap::TCCFeedbackRule` are not a `class` anymore, what are they? 
2. the enum is gone, as enumeration only came with py3.4 that cannot be the problem, as long as there wasn't something else done to enumerate somewhere(somehow) else?

Edit (sorry to spam you): this of course relates to: 
`(midi_map_handle, param, 0, encoder, Live.MidiMap.MapMode.relative_signed_bit, feedback_rule, needs_takeover, sensitivity=1.0)`

**Reply by markusschloesser** [reply]
the stub helpfully lists this:

@staticmethod
    def map_midi_cc_with_feedback_map(midi_map_handle, parameter, midi_channel, controller_number, map_mode, feedback_rule, avoid_takeover, sensitivity=1.0):
        """
        C++ signature :  bool map_midi_cc_with_feedback_map(unsigned int,TPyHandle<ATimeableValue>,int,int,NRemoteMapperTypes::TControllerMapMode,NPythonMidiMap::TCCFeedbackRule,bool [,float=1.0])
        :param midi_map_handle: midi_map_handle
        :type midi_map_handle: int
        :param parameter: parameter
        :type parameter: DeviceParameter
        :param midi_channel: midi_channel
        :type midi_channel: int
        :param controller_number: controller_number
        :type controller_number: int
        :param map_mode: map_mode
        :type map_mode: MapMode
        :param feedback_rule: feedback_rule
        :type feedback_rule: CCFeedbackRule
        :param avoid_takeover: avoid_takeover
        :type avoid_takeover: bool
        :param sensitivity: sensitivity defaults to 1.0 
        :type sensitivity: float
        :rtype: bool
        """
I still don't understand though, of course.

**Reply by BilldarBagdar** [reply]
Is this still relevant after your recent success?

0. The difference between `object` and `NoteFeedbackRule` is that `object` lets a lot more inputs through the front door.  `NoteFeedbackRule` only lets `NoteFeedbackRule` objects through the front door.

1.  I suspect this just means the identifier `class` no longer has meaning in this context so it's not there now, and  `TPyHandle`, `ATimeable` and `NPythonMidiMap::TCCFeedbackRule` are still classes in the same programmatic sense.

What seems like a bigger issue to me is "another change" from Live 10 to 11?  Your first post shows the same 5 input parameters between Live 8 and 10 (more or less).  But your second post shows what I assume is the "equivalent" method signature information about the Live 11 version. (which is great reference, btw)  It looks like the new 11 signature has 3 new input parameters to make 8, and that the `NoteFeedbackRule` 5th parameter "slid down one" from 5th to 6th in the Live 11 parameter list.  Are you seeing what I'm seeing?

2. I suspect "the `enum` is gone" for the same reason as `class` in 1. above.  However, I'm not sure that's what you mean because (coincidentally?) the `NoteFeedbackRule` 5th parameter appears to have "slid down one" to make room for a `MapMode` 5th parameter, and that is the same (missing?) `enum`?

I don't think I understand completely yet, but did any of that help you?

It might be time for you to give me a lesson in "XML/stub" configuration/implementation. :)

**Reply by markusschloesser** [reply]
> It might be time for you to give me a lesson in "XML/stub" configuration/implementation. :)

I just did this (mentioned in pycharm thread):

_"To regenerate the stub for another api version you can just copy the xml file and call it Live.xml then run generate.py. It creates the Live/__init__.py which makes the types understandable for pycharm."_

So you need the Live.xml for Live 10 (I attached one, rename it to Live.xml !) 
[Live10.1.19.xml.zip](https://github.com/markusschloesser/MackieC4_P3/files/6178623/Live10.1.19.xml.zip)
and this script (which I reworked for Py3 but should still work with py2.7)
[AbletonLive-API-Stub-master.zip](https://github.com/markusschloesser/MackieC4_P3/files/6178636/AbletonLive-API-Stub-master.zip)

**Reply by markusschloesser** [reply]
> Is this still relevant after your recent success?

MS still very much relevant, as exactly that error keeps popping up all the time, always related to DeviceParameter. The current theory is, that  `do_add_device_listeners` , `rem_device_listeners`, `add_devicelistener` are the main problem. I am currently trying to rework those (also because I read that the whole `for i in range(len(` thing is rather uncool in Python and should be avoided and a better approach would be to use "enumerate", so instead of doing the above, doing: 

`        for i, tracks in enumerate(tracks):
            self.add_devicelistener(f'{i}: {tracks}', i, type)`
so to not keep track of indexes yourself.
So, as this section doesn't seem to work properly anyway (logically and this seems to be the thing before the error), I tried that, but it now have seem to made a str somewhere, where is shouldn't 🤔 

```
`RemoteScriptError: self.add_device_listeners()

RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Suite\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 759, in add_device_listeners

RemoteScriptError: self.do_add_device_listeners(self.song().tracks, 0)

RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Suite\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 765, in do_add_device_listeners

RemoteScriptError: self.add_devicelistener(f'{i}: {tracks}', i, type)

RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Suite\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 829, in add_devicelistener

RemoteScriptError: track.view.add_selected_device_listener(cb)

RemoteScriptError: AttributeError

RemoteScriptError: 'str' object has no attribute 'view'
`
```
Sorry, I am mixing things in here, which should be in another issue, but fuck it. 😁


> 
> 1. The difference between `object` and `NoteFeedbackRule` is that `object` lets a lot more inputs through the front door.  `NoteFeedbackRule` only lets `NoteFeedbackRule` objects through the front door.

MS: but  `NoteFeedbackRule` is just a bool, how hard can that be? 🙄 See above, I think it's the "DeviceParameter".


> 2. I suspect this just means the identifier `class` no longer has meaning in this context so it's not there now, and  `TPyHandle`, `ATimeable` and `NPythonMidiMap::TCCFeedbackRule` are still classes in the same programmatic sense.

MS: Maybe (?) the log showing `TPyHandle<class ATimeableValue>` for where the "DeviceParameter" should be, is a direct hint? I really don't know, also cos it's showing the same for feedback mode and map type)
 

> 
> What seems like a bigger issue to me is "another change" from Live 10 to 11? Your first post shows the same 5 input parameters between Live 8 and 10 (more or less). But your second post shows what I assume is the "equivalent" method signature information about the Live 11 version. (which is great reference, btw) It looks like the new 11 signature has 3 new input parameters to make 8, and that the `NoteFeedbackRule` 5th parameter "slid down one" from 5th to 6th in the Live 11 parameter list. Are you seeing what I'm seeing?

MS: I revised that table a couple of times yesterday, because I think I accidentally oversaw a comma somewhere, are you looking at the latest version? Cos it seems to me, that the amount stayed the same (8).

> 
> 1. I suspect "the `enum` is gone" for the same reason as `class` in 1. above.  However, I'm not sure that's what you mean because (coincidentally?) the `NoteFeedbackRule` 5th parameter appears to have "slid down one" to make room for a `MapMode` 5th parameter, and that is the same (missing?) `enum`?
> 
> I don't think I understand completely yet, but did any of that help you?

It does, a little bit, thank you! 
But the whole boost/python/c++ signature thing is still my weak point. I am now able to understand when I shouldn't have made a str somewhere out of an int and how to fix this and interpret the log file and boost error, but that's basically it

**Comment by markusschloesser** [comment]
`ATimeableValue` is about automation! I don't remember when that was introduced, but the automation state of a DeviceParameter was only introduced later according to the LOM 8->11 changes
